File: src/cloud/FccMiddleware.Domain/Entities/PreAuthRecord.cs
using FccMiddleware.Domain.Enums;
using FccMiddleware.Domain.Exceptions;
using FccMiddleware.Domain.Common;

namespace FccMiddleware.Domain.Entities;

/// <summary>
/// A pre-authorization request from Odoo POS via an Edge Agent.
/// The idempotency key is (OdooOrderId, SiteCode).
/// Currency amounts are stored as minor units (cents). Volume in millilitres.
/// CustomerTaxId is sensitive — never log this field.
/// Matches pre-auth-record.schema.json.
/// </summary>
public class PreAuthRecord
{
    public Guid Id { get; set; }
    public Guid LegalEntityId { get; set; }
    public string SiteCode { get; set; } = null!;
    public string OdooOrderId { get; set; } = null!;
    public int PumpNumber { get; set; }
    public int NozzleNumber { get; set; }
    public string ProductCode { get; set; } = null!;
    public string CurrencyCode { get; set; } = null!;

    /// <summary>Pre-auth amount in minor units. Always > 0. Authorization is always by amount.</summary>
    public long RequestedAmountMinorUnits { get; set; }

    /// <summary>Price per litre at time of authorization, in minor units per litre. Informational only.</summary>
    public long UnitPriceMinorPerLitre { get; set; }

    public long? AuthorizedAmountMinorUnits { get; set; }
    public long? ActualAmountMinorUnits { get; set; }
    public long? ActualVolumeMillilitres { get; set; }

    /// <summary>actualAmount - requestedAmount in minor units. Negative = under-dispense. Populated at COMPLETED.</summary>
    public long? AmountVarianceMinorUnits { get; set; }

    /// <summary>ABS(amountVariance) / requestedAmount * 10000, rounded. Units: basis points. Populated at COMPLETED.</summary>
    public int? VarianceBps { get; set; }

    public PreAuthStatus Status { get; set; } = PreAuthStatus.PENDING;
    public string? FccCorrelationId { get; set; }
    public string? FccAuthorizationCode { get; set; }
    public string? FailureReason { get; set; }
    public string? VehicleNumber { get; set; }
    [Sensitive]
    public string? CustomerName { get; set; }

    /// <summary>Customer Tax Identification Number (TIN). PII — never log.</summary>
    [Sensitive]
    public string? CustomerTaxId { get; set; }

    public string? CustomerBusinessName { get; set; }
    public string? AttendantId { get; set; }

    /// <summary>
    /// fccTransactionId of the final dispense transaction matched by the reconciliation engine.
    /// Null until COMPLETED state.
    /// </summary>
    public string? MatchedFccTransactionId { get; set; }

    /// <summary>Middleware UUID of the matched Transaction record. Internal FK.</summary>
    public Guid? MatchedTransactionId { get; set; }

    public DateTimeOffset RequestedAt { get; set; }
    public DateTimeOffset ExpiresAt { get; set; }
    public DateTimeOffset? AuthorizedAt { get; set; }
    public DateTimeOffset? DispensingAt { get; set; }
    public DateTimeOffset? CompletedAt { get; set; }
    public DateTimeOffset? CancelledAt { get; set; }
    public DateTimeOffset? ExpiredAt { get; set; }
    public DateTimeOffset? FailedAt { get; set; }
    public int SchemaVersion { get; set; } = 1;
    public DateTimeOffset CreatedAt { get; set; }
    public DateTimeOffset UpdatedAt { get; set; }

    // Navigation properties
    public LegalEntity LegalEntity { get; set; } = null!;

    /// <summary>
    /// Transitions the pre-auth record to <paramref name="newStatus"/> according to the state machine
    /// defined in §5.2 of tier-1-2-state-machine-formal-definitions.md.
    /// Terminal states (no outgoing transitions): COMPLETED, CANCELLED, EXPIRED, FAILED.
    /// Throws <see cref="InvalidPreAuthTransitionException"/> for any invalid transition.
    /// </summary>
    public void Transition(PreAuthStatus newStatus)
    {
        var allowed = (Status, newStatus) switch
        {
            (PreAuthStatus.PENDING,    PreAuthStatus.AUTHORIZED)  => true,
            (PreAuthStatus.PENDING,    PreAuthStatus.CANCELLED)   => true,
            (PreAuthStatus.PENDING,    PreAuthStatus.EXPIRED)     => true,
            (PreAuthStatus.PENDING,    PreAuthStatus.FAILED)      => true,
            (PreAuthStatus.AUTHORIZED, PreAuthStatus.DISPENSING)  => true,
            (PreAuthStatus.AUTHORIZED, PreAuthStatus.COMPLETED)   => true,
            (PreAuthStatus.AUTHORIZED, PreAuthStatus.CANCELLED)   => true,
            (PreAuthStatus.AUTHORIZED, PreAuthStatus.EXPIRED)     => true,
            (PreAuthStatus.AUTHORIZED, PreAuthStatus.FAILED)      => true,
            (PreAuthStatus.DISPENSING, PreAuthStatus.COMPLETED)   => true,
            (PreAuthStatus.DISPENSING, PreAuthStatus.CANCELLED)   => true,
            (PreAuthStatus.DISPENSING, PreAuthStatus.EXPIRED)     => true,
            (PreAuthStatus.DISPENSING, PreAuthStatus.FAILED)      => true,
            _ => false
        };

        if (!allowed)
            throw new InvalidPreAuthTransitionException(Status, newStatus);

        Status = newStatus;
        UpdatedAt = DateTimeOffset.UtcNow;
    }
}

File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/buffer/entity/PreAuthRecord.kt
package com.fccmiddleware.edge.buffer.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import com.fccmiddleware.edge.adapter.common.PreAuthStatus

/**
 * Pre-authorization record stored locally on the Edge Agent.
 *
 * Pre-auth is the top latency path: POST /api/preauth must respond
 * based on LAN-only work. Cloud forwarding is always async and never
 * on the request path.
 *
 * All timestamps: ISO 8601 UTC TEXT. Money: Long minor units. UUIDs: TEXT.
 * customer_tax_id is sensitive — NEVER log this field.
 */
@Entity(
    tableName = "pre_auth_records",
    indices = [
        // Idempotency: prevent duplicate pre-auth creation from Odoo retries
        Index(value = ["odoo_order_id", "site_code"], unique = true, name = "ix_par_idemp"),
        // Cloud forward worker: find unsynced records ordered by time
        Index(value = ["is_cloud_synced", "created_at"], name = "ix_par_unsent"),
        // Expiry worker: find active records approaching or past expiry
        Index(value = ["status", "expires_at"], name = "ix_par_expiry"),
    ]
)
data class PreAuthRecord(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "site_code")
    val siteCode: String,

    @ColumnInfo(name = "odoo_order_id")
    val odooOrderId: String,

    /** FCC pump number (after Odoo → FCC translation via Nozzle table) */
    @ColumnInfo(name = "pump_number")
    val pumpNumber: Int,

    /** FCC nozzle number */
    @ColumnInfo(name = "nozzle_number")
    val nozzleNumber: Int,

    @ColumnInfo(name = "product_code")
    val productCode: String,

    @ColumnInfo(name = "currency_code")
    val currencyCode: String,

    /** Minor units (cents). NEVER floating point. */
    @ColumnInfo(name = "requested_amount_minor_units")
    val requestedAmountMinorUnits: Long,

    /** Minor units per litre. Null for legacy rows created before unit price was persisted. */
    @ColumnInfo(name = "unit_price_minor_per_litre")
    val unitPrice: Long? = null,

    /** Minor units. Null until FCC authorizes. */
    @ColumnInfo(name = "authorized_amount_minor_units")
    val authorizedAmountMinorUnits: Long?,

    /** PreAuthStatus: PENDING | AUTHORIZED | DISPENSING | COMPLETED | EXPIRED | CANCELLED | FAILED */
    @ColumnInfo(name = "status")
    val status: PreAuthStatus = PreAuthStatus.PENDING,

    @ColumnInfo(name = "fcc_correlation_id")
    val fccCorrelationId: String?,

    @ColumnInfo(name = "fcc_authorization_code")
    val fccAuthorizationCode: String?,

    @ColumnInfo(name = "failure_reason")
    val failureReason: String?,

    @ColumnInfo(name = "customer_name")
    val customerName: String?,

    /** SENSITIVE — never log */
    @ColumnInfo(name = "customer_tax_id")
    val customerTaxId: String?,

    @ColumnInfo(name = "raw_fcc_response")
    val rawFccResponse: String?,

    /** ISO 8601 UTC */
    @ColumnInfo(name = "requested_at")
    val requestedAt: String,

    /** ISO 8601 UTC; null until FCC authorizes */
    @ColumnInfo(name = "authorized_at")
    val authorizedAt: String?,

    /** ISO 8601 UTC; null until dispense completes */
    @ColumnInfo(name = "completed_at")
    val completedAt: String?,

    /** ISO 8601 UTC; pre-auth TTL from site config */
    @ColumnInfo(name = "expires_at")
    val expiresAt: String,

    /** Boolean: 0/1 */
    @ColumnInfo(name = "is_cloud_synced")
    val isCloudSynced: Int = 0,

    @ColumnInfo(name = "cloud_sync_attempts")
    val cloudSyncAttempts: Int = 0,

    /** ISO 8601 UTC; null until first cloud sync attempt */
    @ColumnInfo(name = "last_cloud_sync_attempt_at")
    val lastCloudSyncAttemptAt: String?,

    @ColumnInfo(name = "schema_version")
    val schemaVersion: Int = 1,

    /** ISO 8601 UTC */
    @ColumnInfo(name = "created_at")
    val createdAt: String,
)
